src/thor/route_matcher.cc

```diff
@@ -427,20 +427,40 @@ bool RouteMatcher::FormPath(const sif::mode_costing_t& mode_costing,
         if (expand_from_node(mode_costing, mode, reader, options.shape(), distances, time_info,
                              options.use_timestamps(), index, end_node_tile, de->endnode(), end_nodes,
                              prev_edge_label, elapsed, path_infos, false, end_node, followed_edges)) {
-          // If node equals stop node then when are done expanding - get the matching end edge
+          // Find the edge we stopped on at the destination, if we didnt find it the greedy algorithm
+          // hit a local maximum (made the wrong choice), TODO: we could rollback and try more
           auto n = end_nodes.find(end_node);
           if (n == end_nodes.end()) {
             return false;
           }
 
+          // When the route ends at a node in the graph we have an ambiguous case. Multiple
+          // destination edge candidates could have ended at this node but we use an unordered_map
+          // instead of a multimap, which means when we go to insert the other candidates that end
+          // there, they dont get inserted, only the first one does. This is all we really need for
+          // finding the path but it means that once we do find the path to that node, the edge that
+          // was in the value portion of the map entry might be the wrong edge. So here we need to
+          // go find the edge candidate that was actually used in the path
+          auto found_edge = destination.path_edges().end();
+          if (n->second.first.end_node()) {
+            found_edge =
+                std::find_if(destination.path_edges().begin(), destination.path_edges().end(),
+                             [&path_infos](const auto& e) -> bool {
+                               return e.graph_id() == path_infos.back().edgeid;
+                             });
+            if (found_edge == destination.path_edges().end()) {
+              throw std::logic_error("Could not find destination candidate in shape-walked path");
+            }
+          }
+
           // TODO: when we actually have more than one leg, we have to do when we break legs
           // Store the matching edge candidates in the shapes locations
-          const auto& end_edge = n->second.first;
-          options.mutable_shape(0)->mutable_path_edges()->Add()->CopyFrom(end_edge);
```

Comment:
this was a copy-paste-fail which caused the origin edge candidate to pretty much always be wrong except for in trivial routes.